[ONNX] Support more sympy operations in fx-onnx exporter

titaiwangms opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #112758

Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

~~NOTE: `native_batch_norm` regression is caused by https://github.com/microsoft/onnxscript/issues/1140. Will fix it before I merge this.~~

titaiwangms committed:
[ONNX] Support more sympy operations in fx-onnx exporter

[ghstack-poisoned]

thiagocrepaldi commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -129,6 +131,59 @@ def func(x, b=torch.tensor(1.0)):
         )
 
+    @pytorch_test_common.skip_dynamic_fx_test(
+        "sympy operation tests doesn't need dynamic shape"
```
```suggestion
        "sympy operation tests don't need dynamic shape"
```
NIT: SPELLING

thiagocrepaldi commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -129,6 +131,59 @@ def func(x, b=torch.tensor(1.0)):
+        )
+
+    @pytorch_test_common.skip_dynamic_fx_test(
+        "sympy operation tests doesn't need dynamic shape"
```
```suggestion
        "sympy operation tests don't need dynamic shape"
```

thiagocrepaldi commented on test/onnx/test_fx_to_onnx_with_onnxruntime.py:
```diff
@@ -129,6 +131,59 @@ def func(x, b=torch.tensor(1.0)):
+    def test_sympy_operatons_return_numeric(self):
+        def func(x, y):
+            return (
+                # torch.tensor([operator.add(x.item(), y.item())]),
+                # torch.tensor([operator.sub(x.item(), y.item())]),
+                # torch.tensor([operator.mul(x.item(), y.item())]),
+                # torch.tensor([operator.truediv(x.item(), y.item())]),
+                # torch.tensor([operator.floordiv(x.item(), y.item())]),
+                # torch.tensor([operator.pow(x.item(), y.item())]),
+                # torch.tensor([operator.abs(x.item())]),
+                # torch.tensor([operator.neg(x.item())]),
```
why all these comments? are they from previous testing?

either delete them or add an additional explanation for them

thiagocrepaldi commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -240,12 +243,17 @@ def _fill_tensor_shape_type(
-            onnxscript_value.dtype = fx_type_utils.from_sym_value_to_torch_dtype(
-                expected_value
-            )
+            if node.target == math.ceil or node.target == math.floor:
+                # Type promotion for ceil/floor from built-in functions.
+                # ceil and floor have float outputs in torchlib
+                onnxscript_value.dtype = torch.float32
```
should this be part of the type promotion logic? adding hard-coded op-specific logic in the dispatcher makes it hard to maintain

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"




[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"




[ghstack-poisoned]

thiagocrepaldi commented on torch/onnx/_internal/fx/registration.py:
```diff
@@ -68,6 +69,15 @@ def from_qualified_name(cls, qualified_name: str) -> OpName:
+    def from_builtin_function(
+        cls, builtin_function: types.BuiltinFunctionType
+    ) -> OpName:
+        op = builtin_function.__name__  # add, sub, etc.
```
add docstring

thiagocrepaldi reviewed (commented):
Awesome, having built-in ops seems a much more robust solution

thiagocrepaldi commented:
NIT: did you really mean "sympy" in the PR title? or was it "built-in"?

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

NOTE: `native_batch_norm` regression is caused by https://github.com/microsoft/onnxscript/issues/1140. Will fix it before I merge this.

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge

justinchuby commented on torch/onnx/_internal/fx/fx_onnx_interpreter.py:
```diff
@@ -147,9 +148,9 @@ def _retrieve_or_adapt_input_to_graph_set(
-            output = onnxscript.opset18.Concat(*sequence_mixed_elements, axis=0)
-        output.dtype = torch.int64
-        output.shape = [len(sequence_mixed_elements)]
+            output = onnxscript.opset18.Concat(*sequence_mixed_elements, axis=0)  # type: ignore[type-var]
```
We should alias the concat op in torch lib in the near future so this call is opset version agnostic

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

~~NOTE: `native_batch_norm` regression is caused by https://github.com/microsoft/onnxscript/issues/1140. Will fix it before I merge this.~~

[ghstack-poisoned]

titaiwangms committed:
Update on "[ONNX] Support more sympy operations in fx-onnx exporter"


Fix https://github.com/microsoft/onnx-converters-private/issues/190

This PR retires built-in function mapping by adding built-in ops into torchlib (https://github.com/microsoft/onnxscript/pull/1135), and provide a runtime tests to guard the operation conversion.

More built-in ops are supported in torchlib as well.

~~NOTE: `native_batch_norm` regression is caused by https://github.com/microsoft/onnxscript/issues/1140. Will fix it before I merge this.~~

[ghstack-poisoned]

titaiwangms commented:
@pytorchbot merge
